ramp4-pcar4/ramp4-pcar4
# Fix Max WebGL Context Issue

### Related Item(s)
For #1908 

### Notes
- Most browsers have a webgl context limit of 16 (one instance of RAMP can use up to two instances - one for the overall map and another for the overview)
- Once the number of contexts on a page is higher than the limit, the oldest existing context will be lost. This previously resulted in all RAMP instances on the page breaking instead of only the instance that lost its context
- This solution does not require any changes to code for products that use RAMP (like Storylines)

### Changes
- Now when the oldest existing context is lost for an instance, all other RAMP instances will still be functional
- When a RAMP instance is visible on the screen, if its overview or map context is lost, it will be recovered

### Testing
There is a new demo in the catalogue labelled "Multiple RAMP instances inside the Wet template". It contains 9 RAMP instances that only render once visible on the screen. Was trying to stimulate something similar to viewing RAMP maps in Storylines and producing the notorious console warning `Too many active WebGL contexts. Oldest context will be lost`, so it can be removed once this issue is fully finished if we want to.

1. Open the demo and console. Slowly scroll down, rendering and keeping track of the number of webgl contexts used. Each instance rendered is two more webgl contexts
2. Once you reach the 8th instance, 16 contexts exist and we have reached the limit. Everything should be fine so far.
3. When the 9th instance renders, the warning should appear in the console. At this point we have 18 contexts, so two contexts should be lost. Following the order, the oldest two contexts should be the first instances map and overview. 
4. Scroll back to the first instance. Feel free to test the other instance on your way back to the top, they should still be functional. If you scroll back to the first instance you should briefly see that both the map and overview are empty (white background), and will reappear shortly once visible on the screen
5. The warning will appear as we are at 18 contexts again, and the oldest contexts will be lost, which will be the second instances map and overview. Once the second instance is visible, the contexts will be recovered and the cycle continues.

Note that if you render the instances in a different order, or are scrolling all over the place, there can be cases where an instance can have its overview context lost but still have its map context or vice versa. In these cases the respective missing context will reappear as well. No matter what happens when an instance is visible it should always have both its contexts.

Changes were also tested in Storylines with RAMP4, and similar behaviour was seen there. All maps were functional and once you viewed the instance with a lost context it was recovered.


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2041)
